tidy3d/plugins/resonance/resonance.py

```diff
@@ -0,0 +1,272 @@
+        """Solve a generalized eigenvalue problem of the form
+        Av = lambda B v. We make B nonsingular by eliminating eigenvalues
+        smaller than rcond times the largest. This method is used in
+        Michael R. Wall and Daniel Neuhauser,
+        "Extraction, through filter-diagonalization, of general quantum eigenvalues
+        or classical normal mode frequencies of a short-time segment of a signal.
+        I. Theory and application to a quantum-dynamics model,"
+        J. Chem. Phys. 102, 8001 (1995).
+
+        """
+        eigvals_b, eigvecs_b = scipy.linalg.eig(b_matrix)
+        large_inds = np.abs(eigvals_b) > rcond * np.amax(np.abs(eigvals_b))
+        eigvals_b = eigvals_b[large_inds]
+        eigvecs_b = eigvecs_b[:, large_inds]
+        eigvecs_b = self._normalize_eigvecs(eigvecs_b)
+
+        change_basis = eigvecs_b @ np.diag(1 / np.sqrt(eigvals_b))
+        new_a_matrix = change_basis.T @ a_matrix @ change_basis
+        eigvals, eigvecs_a = scipy.linalg.eig(new_a_matrix)
+        eigvecs_a = self._normalize_eigvecs(eigvecs_a)
+
+        eigvecs = change_basis @ eigvecs_a
+
+        return eigvals, eigvecs
+
+    def _find_amplitudes(
+        self,
+        signal: np.ndarray[complex],
+        prev_eigvals: np.ndarray[complex],
+        eigvals: np.ndarray[complex],
+        eigvecs: np.ndarray[complex],
+    ) -> np.ndarray[complex]:
+        """Compute the amplitudes"""
+        M = int(len(signal) / 2) - 2
+        J = len(eigvals)
+        zvals = prev_eigvals / np.abs(prev_eigvals)
+        amplitudes = np.zeros(J, dtype=complex)
+
+        zinvl = np.exp(1j * np.outer(np.log(zvals) * 1j, np.arange(M + 1)))
+        factor = zinvl @ signal[: M + 1]
+
+        for k in range(J):
+            amplitudes[k] = np.square(eigvecs[:, k] @ factor)
+
+        return amplitudes
+
+    def _find_errors(
+        self,
+        eigvals: np.ndarray[complex],
+        u_matrices: np.ndarray[complex],
+        eigvecs: np.ndarray[complex],
+    ) -> np.ndarray[complex]:
+        """Find the eigenvalue error"""
+        J = len(eigvals)
+        errors = np.zeros(J)
+        for k in range(J):
+            errors[k] = np.linalg.norm(
+                (u_matrices[2] - (eigvals[k] ** 2) * u_matrices[0]) @ eigvecs[:, k]
+            )
+        return errors
+
+    def _iterate(
+        self, signal: np.ndarray[complex], prev_eigvals: np.ndarray[complex]
+    ) -> ResonanceData:
+        """Run a single iteration of the solver."""
+        u_matrices = self._evaluate_matrices(signal, prev_eigvals)
+
+        eigvals, eigvecs = self._solve_gen_eig_prob(u_matrices[1], u_matrices[0], self.rcond)
+
+        errors = self._find_errors(eigvals, u_matrices, eigvecs)
+
+        amplitudes = self._find_amplitudes(signal, prev_eigvals, eigvals, eigvecs)
+
+        return ResonanceData(eigvals=eigvals, errors=errors, complex_amplitudes=amplitudes)
+
+    def run(self, signal: np.ndarray[complex], time_step: float) -> xr.Dataset:
```

Comment:
can we put these two public facing methods at the top of the class? just about` _get_resonance_info`? More likely users will be coming in to find these methods so will make them easier to locate.